mono/metadata/class.c

```diff
@@ -6117,42 +6128,84 @@ mono_generic_class_get_class (MonoGenericClass *gclass)
 	return klass;
 }
 
+static MonoImage *
+get_image_for_container (MonoGenericContainer *container)
+{
+	MonoImage *result;
+	if (container->is_anonymous) {
+		result = container->owner.image;
+	} else {
+		MonoClass *klass;
+		if (container->is_method) {
+			MonoMethod *method = container->owner.method;
+			g_assert_checked (method);
+			klass = method->klass;
+		} else {
+			klass = container->owner.klass;
+		}
+		g_assert_checked (klass);
+		result = klass->image;
+	}
+	g_assert (result);
+	return result;
+}
+
+MonoImage *
+get_image_for_generic_param (MonoGenericParam *param)
+{
+	MonoGenericContainer *container = mono_generic_param_owner (param);
+	g_assert_checked (container);
+	return get_image_for_container (container);
+}
+
+// Make a string in the designated image consisting of a single integer.
+#define INT_STRING_SIZE 16
+char *
+make_generic_name_string (MonoImage *image, int num)
+{
+	char *name = mono_image_alloc0 (image, INT_STRING_SIZE);
+	snprintf (name, INT_STRING_SIZE, "%d", num);
+	return name;
+}
+
+// This is called by mono_class_from_generic_parameter_internal when a new class must be created.
+// pinfo is derived from param by the caller for us.
 static MonoClass*
-make_generic_param_class (MonoGenericParam *param, MonoImage *image, gboolean is_mvar, MonoGenericParamInfo *pinfo)
+make_generic_param_class (MonoGenericParam *param, MonoGenericParamInfo *pinfo)
 {
 	MonoClass *klass, **ptr;
 	int count, pos, i;
 	MonoGenericContainer *container = mono_generic_param_owner (param);
+	g_assert_checked (container);
 
-	if (!image)
-		/* FIXME: */
-		image = mono_defaults.corlib;
+	MonoImage *image = get_image_for_container (container);
+	gboolean is_mvar = container->is_method;
+	gboolean is_anonymous = container->is_anonymous;
 
 	klass = mono_image_alloc0 (image, sizeof (MonoClass));
 	classes_size += sizeof (MonoClass);
 
 	if (pinfo) {
-		klass->name = pinfo->name;
+		CHECKED_METADATA_WRITE_PTR_EXEMPT ( klass->name, pinfo->name );
```

Comment:
Mixing adding the checks with the core change makes reviewing harder.

It makes for worse commit history as it make the same commit show up in more places than needed   when checking log or blame for a file.